### ISSUE TITLE
[Milestone 5] Post-FIRE Advance：消除 Angle Pinning

### ISSUE THREAD

**Issue by Firestl** [Firestl/gold-miner-sim#9]
## 目标

Milestone 4 使用 `SwingDecisionWrapper` 后，DQN 在 random map 上首次超过 random baseline，但出现稳定的 **angle pinning**：FIRE 完成出钩/回收后，hook 回到原发射角并立即重新获得决策权，策略容易在同一角度反复 FIRE。

本阶段只修改 **FIRE 后的下一次决策时机**：

- `WAIT` 语义保持不变：推进 10 个 physics ticks；
- `FIRE` 先自动完成完整出钩/回收；
- 如果 episode 未结束且 hook 已回到 `SWINGING`，再自动推进 10 个 `WAIT` physics ticks；
- 然后才把 observation 返回给 Agent。

保持 random map、26 维 observation、reward、action space、physics、spawn points 和 DQN 配置不变。

目标是消除“回收结束后在同一角度立即再次 FIRE”的结构性循环，并观察 DQN 是否形成更稳定的地图相关出钩策略。

---

## 1. 不修改底层 `GoldMinerEnv`

本 Issue 不修改：

- `GoldMinerEnv`；
- `fixed/random` map mode；
- 12 个 random spawn points；
- action space：`Discrete(2)`；
- `WAIT=0` / `FIRE=1`；
- observation：26 维 `float32`；
- reward：`new_score - old_score`；
- episode duration：60 s；
- physics：60 Hz；
- Hook 状态机；
- collision；
- object radius / value / retract speed；
- Pygame renderer。

不要通过修改环境或 reward 改善训练结果。

---

## 2. 保留现有 wrapper

以下 wrapper 全部保留原行为：

```text
DecisionIntervalWrapper
SwingDecisionWrapper
```

不要修改其现有语义和测试。

新增一个 wrapper：

```python
SwingAdvanceDecisionWrapper
```

用途：作为 Milestone 5 的训练/评估 wrapper，与 Milestone 4 的 `SwingDecisionWrapper` 做可复现实验对照。

---

## 3. 新增 `SwingAdvanceDecisionWrapper`

修改：

```text
src/gold_miner_sim/wrappers.py
```

新增：

```python
class SwingAdvanceDecisionWrapper(gymnasium.Wrapper):
    ...
```

固定常量：

```text
ADVANCE_INTERVAL = 10 physics ticks
```

不要做成 CLI、配置文件或超参数。

正常未结束 episode 时，每次 `step()` 返回时底层 hook 必须处于：

```text
SWINGING
```

---

## 4. Action 校验

wrapper 必须保持底层 action space 契约。

使用：

```python
if not self.action_space.contains(action):
    raise ValueError(...)
```

不要使用仅比较值是否等于 `0/1` 的方式。

必须接受合法的 Python / NumPy integer action；必须拒绝：

```text
-1
2
0.0
1.0
其他不属于 Discrete(2) 的值
```

---

## 5. `WAIT` 精确语义

Agent 在 `SWINGING` 时选择：

```text
WAIT
```

执行最多 10 个底层 tick：

```text
WAIT x 10
```

要求：

1. 每 tick 调用 `env.step(WAIT)`；
2. 累计 reward；
3. 任意 tick `terminated/truncated` 时立即返回；
4. 正常情况下执行满 10 ticks；
5. 返回最后一个 tick 的 observation / info / flags；
6. 返回累计 reward。

该行为必须与：

```text
SwingDecisionWrapper.step(WAIT)
```

以及裸环境手工 `WAIT x 10` 等价。

---

## 6. `FIRE` 精确语义

Agent 在 `SWINGING` 时选择：

```text
FIRE
```

一次 `SwingAdvanceDecisionWrapper.step(FIRE)` 分为两个阶段。

### 6.1 Phase A：完整 FIRE cycle

先执行：

```text
env.step(FIRE)
```

之后只执行：

```text
env.step(WAIT)
```

直到满足任一条件：

1. hook 重新进入 `SWINGING`；
2. `terminated=True`；
3. `truncated=True`。

整个阶段累计 reward。

不得在首次回到 `SWINGING` 后多执行一个 tick。

这部分语义应与 `SwingDecisionWrapper.step(FIRE)` 的 FIRE cycle 一致。

### 6.2 Phase B：Post-FIRE advance

如果 Phase A 结束时 episode 已结束：

```text
立即返回
```

不要执行 post-FIRE advance。

如果 episode 未结束且 hook 已回到 `SWINGING`：

继续执行最多：

```text
WAIT x 10
```

即：

```text
ADVANCE_INTERVAL = 10
```

每 tick：

- 调用 `env.step(WAIT)`；
- 累计 reward；
- 检查 `terminated/truncated`；
- episode 结束则立即返回。

正常执行满 10 ticks 后再返回。

### 6.3 最终返回

返回：

```text
最后一个底层 tick 的 observation
累计 reward
terminated
truncated
最后一个底层 tick 的 info
```

正常情况下：

```text
FIRE at angle A
    -> extend/retract
    -> back to SWINGING at angle A
    -> WAIT x 10
    -> return at next decision angle
```

因此下一次决策 observation 不应仍停留在原发射角。

---

## 7. Reward 不修改

底层 reward 仍然是：

```text
reward = new_score - old_score
```

wrapper 只累计执行期间底层 reward。

例如成功收回 DIAMOND：

```text
FIRE
-> hit
-> retract
-> +500
-> SWINGING
-> WAIT x 10
-> return
```

该 wrapper step 返回：

```text
reward = 500
```

不要新增：

- hit reward；
- FIRE penalty/reward；
- empty-shot penalty；
- WAIT penalty；
- angle reward；
- proximity reward。

---

## 8. Angle pinning 的结构性验收

本 Issue 的核心行为必须有明确 regression test。

选择一个确定不会 timeout 的 FIRE 场景，记录：

```text
fire_angle = angle before step(FIRE)
```

调用：

```text
wrapper.step(FIRE)
```

正常返回后必须满足：

```text
hook_state == SWINGING
returned_angle != fire_angle
```

在不经过边界特殊情况的普通场景中，当前物理参数下应约推进 10°。

例如：

```text
fire_angle = +20°
```

若 swing direction 向右，正常返回应约为：

```text
+30°
```

不要把“必须 +10°”写成对所有角度成立的规则，因为 ±70° 边界会发生反向；测试应覆盖一个非边界场景。

---

## 9. 边界行为

Post-FIRE advance 必须直接复用底层 swing physics。

不要自行计算下一角度，不要直接写：

```python
angle += 10
```

例如发射结束时 hook 位于接近 `+70°` 且继续向右，执行 WAIT 后应按底层逻辑：

```text
到 +70°
反向
向左继续摆动
```

wrapper 只负责调用 `env.step(WAIT)` 10 次。

---

## 10. Timeout 语义

必须分别处理两种 timeout。

### A. FIRE cycle 内 timeout

如果在：

```text
EXTENDING / RETRACT_*
```

期间时间到：

- 立即返回；
- 不执行 post-FIRE WAIT；
- `truncated=True`；
- `remaining_time=0`；
- 未完成回收的物体不加分。

### B. Post-FIRE advance 内 timeout

如果 FIRE 已完整结束并重新 SWINGING，但之后的 `WAIT x 10` 中时间到：

- 在第一个 `truncated=True` tick 立即返回；
- 不执行剩余 WAIT；
- 返回实际累计 reward；
- `remaining_time=0`。

---

## 11. Training 切换到新 wrapper

修改：

```text
scripts/train_dqn.py
```

环境链改为：

```text
GoldMinerEnv(render_mode=None, map_mode="random")
    -> SwingAdvanceDecisionWrapper
    -> Monitor
    -> DQN
```

### DQN 配置保持不变

| 参数 | 值 |
|---|---|
| algorithm | `DQN` |
| policy | `MlpPolicy` |
| total_timesteps | `200_000` |
| seed | `0` |
| buffer_size | `100_000` |
| learning_starts | `1_000` |
| exploration_fraction | `0.3` |
| 其他参数 | SB3 默认 |

不要修改：

- learning rate；
- gamma；
- exploration 参数；
- batch size；
- target update；
- train frequency；
- network architecture。

默认输出改为：

```text
models/dqn_gold_miner_advance.zip
runs/dqn_advance/
```

`--seed` 仍然是整个训练实验 seed，用于 SB3 和 environment/map RNG 的可复现初始化。

---

## 12. Random baseline 使用新 wrapper

修改：

```text
scripts/random_baseline.py
```

环境链改为：

```text
GoldMinerEnv(render_mode=None, map_mode="random")
    -> SwingAdvanceDecisionWrapper
```

默认仍然：

```text
episodes = 100
map seeds = 1000..1099
```

第 `i` 局：

```text
map_seed = 1000 + i
```

Action-space RNG 整次 baseline 只 seed 一次。

输出保持：

```text
Episodes
Seed range
Mean score
Std score
Min score
Max score
```

Milestone 4 的 random baseline `545.5` 不能直接用于本阶段比较，必须重新运行。

---

## 13. DQN evaluation 使用新 wrapper

修改：

```text
scripts/eval_dqn.py
```

环境链改为：

```text
GoldMinerEnv(map_mode="random")
    -> SwingAdvanceDecisionWrapper
```

Headless 默认：

```text
100 episodes
seeds 1000..1099
```

模型：

```text
models/dqn_gold_miner_advance.zip
```

Inference：

```python
model.predict(obs, deterministic=True)
```

输出保持：

```text
Episodes
Seed range
Mean score
Std score
Min score
Max score
```

继续支持：

```text
--render --seed <map_seed>
```

render 模式默认 1 episode。

---

## 14. 最低测试

不要在 pytest 中训练 DQN。

新增 `SwingAdvanceDecisionWrapper` 测试，至少覆盖：

### 14.1 WAIT 等价性

```text
wrapper.step(WAIT)
```

等价于裸环境：

```text
WAIT x 10
```

比较 observation / reward / angle / hook state / rope length / remaining time。

### 14.2 Empty FIRE + advance

构造确定空钩场景。

一次：

```text
wrapper.step(FIRE)
```

必须完成：

```text
EXTENDING
-> RETRACT_EMPTY
-> SWINGING
-> WAIT x 10
-> return
```

验证：

- reward=0；
- hook state=SWINGING；
- rope=min；
- angle 已离开发射角；
- tick 数等于 FIRE cycle ticks + 10。

### 14.3 Loaded FIRE + advance

构造确定命中场景。

一次 FIRE 后验证：

- object 已收回；
- reward 等于 value；
- score 已增加；
- object inactive；
- attached object=None；
- hook state=SWINGING；
- 回收完成后又推进 10 WAIT ticks；
- 返回角度不是发射角。

### 14.4 Angle pinning regression

选择非边界发射角。

必须明确断言：

```text
returned_angle != fire_angle
```

并验证返回 angle 与裸环境执行：

```text
完整 FIRE cycle + WAIT x 10
```

完全一致。

### 14.5 Boundary reflection

构造接近 `+70°` 或 `-70°` 的 FIRE 场景。

FIRE cycle 后的 post-FIRE WAIT 必须使用底层 swing physics 正常反向，wrapper 不得手工修改 angle。

### 14.6 Timeout during FIRE cycle

验证：

- 立即 truncated；
- 不执行 post-FIRE advance；
- 未完成回收不计分。

### 14.7 Timeout during post-FIRE advance

构造 FIRE cycle 已结束、但剩余时间不足 10 ticks 的场景。

验证：

- 在 timeout tick 立即返回；
- `_ticks` 最终为 episode 上限；
- 不执行剩余 ticks。

### 14.8 Episode reward consistency

完整 episode：

```text
sum(wrapper rewards) == final env score
```

### 14.9 旧 wrapper 回归

现有：

```text
DecisionIntervalWrapper
SwingDecisionWrapper
```

全部测试继续通过。

---

### A. Tests

```bash
uv run pytest -q
```

必须全部通过。

### B. Random baseline

```bash
uv run --group train python scripts/random_baseline.py --episodes 100 --seed 1000
```

记录：

```text
mean_random_advance
std
min
max
```

### C. Train

```bash
uv run --group train python scripts/train_dqn.py --timesteps 200000 --seed 0
```

必须生成：

```text
models/dqn_gold_miner_advance.zip
runs/dqn_advance/
```

同时记录 Monitor 中 episode 数量。

### D. Headless evaluation

```bash
uv run --group train python scripts/eval_dqn.py --model models/dqn_gold_miner_advance.zip --episodes 100 --seed 1000
```

记录：

```text
mean_dqn_advance
std
min
max
```

### E. Visual / behavior evaluation

至少回放：

```text
seed 1000
seed 1007
seed 1042
```

重点观察并在 PR 描述中记录：

- 是否还出现同一 angle 连续反复 FIRE；
- FIRE angle 是否随地图变化；
- 是否会扫描多个决策角度；
- 是否仍明显忽略 GOLD。

---

## 16. 实验结果记录

PR 描述至少记录：

```text
Tests
Random baseline
DQN training timesteps
Training seed
Monitor episode count
DQN evaluation
```

并给出：

```text
mean_random_advance
mean_dqn_advance
delta = mean_dqn_advance - mean_random_advance
```

同时列出 Milestone 4 结果作为参考：

```text
M4 SwingDecisionWrapper:
random = 545.5
dqn    = 571.0
delta  = +25.5
```

注意：不同 wrapper 下相同 `200_000` SB3 timesteps 对应的实际 physics exposure / episode 数量可能不同。因此 M4/M5 的 absolute score 对比只能作为行为参考，不要声称这是严格固定 physics budget 的算法性能提升。

---

## 17. 学习效果判断

希望观察：

```text
mean_dqn_advance > mean_random_advance
```

更重要的是行为上：

```text
同一角度连续 FIRE 循环消失
```

不同地图下 FIRE angle pattern 应有明显变化。

以上属于实验目标，不进入 pytest，也不是代码合并硬门槛。

如果 DQN 没有超过 random：

- 不调超参数；
- 不延长默认训练；
- 不做 reward shaping；
- 不换 PPO；
- 在 PR 中记录实际结果和行为现象。

---

## 18. 明确不做

本 Issue 不实现：

- PPO / A2C / 其他算法；
- DQN 超参数搜索；
- reward shaping；
- duration-aware discount / SMDP；
- n-step return；
- prioritized replay；
- 自定义 replay buffer；
- 自定义 policy network；
- 修改 observation；
- 修改 random spawn points；
- 连续随机地图生成；
- vectorized env；
- TensorBoard / W&B / MLflow；
- RGB / YOLO；
- 新物体、炸药、商店；
- GOLD 价值/时间权衡专项优化。

---

## Definition of Done

- [ ] 新增 `SwingAdvanceDecisionWrapper`；
- [ ] 保留 `DecisionIntervalWrapper` 和 `SwingDecisionWrapper` 原行为；
- [ ] WAIT 最多推进 10 physics ticks；
- [ ] FIRE 完整自动推进出钩/回收；
- [ ] FIRE cycle 正常结束后自动推进额外 10 个 SWINGING WAIT ticks；
- [ ] FIRE 后正常返回 angle 不再停留在原发射角；
- [ ] action validation 使用 `action_space.contains()`；
- [ ] FIRE cycle timeout 时不执行 post-FIRE advance；
- [ ] post-FIRE advance timeout 时立即停止；
- [ ] reward 聚合正确；
- [ ] episode reward 总和等于最终 score；
- [ ] 新 wrapper 有上述最低测试覆盖；
- [ ] 现有全部测试继续通过；
- [ ] training / baseline / evaluation 全部切换到新 wrapper；
- [ ] 默认模型输出为 `models/dqn_gold_miner_advance.zip`；
- [ ] Monitor 日志为 `runs/dqn_advance/`；
- [ ] 完成一次默认 200k DQN 训练；
- [ ] baseline / DQN 使用相同 evaluation seeds 1000..1099；
- [ ] PR 记录 baseline、training、evaluation 和 delta；
- [ ] PR 记录 angle-pinning 是否消失及至少 3 个 seed 的行为观察；
- [ ] 未实现本 Issue 明确排除的内容。
